flink-contrib/flink-timerserivce-rocksdb/src/main/java/org/apache/flink/contrib/streaming/api/operators/RocksDBInternalTimerService.java

```diff
@@ -0,0 +1,797 @@
+
+		try {
+			int keyGroup = IntSerializer.INSTANCE.deserialize(inputView);
+			long timestamp = LongSerializer.INSTANCE.deserialize(inputView);
+			K key = keySerializer.deserialize(inputView);
+			N namespace = namespaceSerializer.deserialize(inputView);
+
+			return new Tuple4<>(keyGroup, timestamp, key, namespace);
+		} catch (IOException e) {
+			throw new RuntimeException("Error while deserializing the raw timer.", e);
+		}
+	}
+
+	/**
+	 * A timer store backed by RocksDB. 
+	 * 
+	 * The timers are stored in RocksDB in the order of key groups. To allow 
+	 * efficient access, the timers are partitioned and the leader of each 
+	 * partition is stored in an in-memory heap. The top of the heap is 
+	 * exactly the first timer to trigger. The heap is updated whenever the 
+	 * partition's leader is updated.
+	 */
+	private class PersistentTimerHeap {
+
+		private final ColumnFamilyHandle handle;
+
+		/** Leader timers in the partitions */
+		private final Tuple4<Integer, Long, K, N>[] partitionHeadTimers;
+		
+		/** The order of the partition's leader in the heap */
+		private final int[] orders;
+		
+		/** The partitions in the order of their leaders in the heap */
+		private final int[] indices;
+		
+		@SuppressWarnings("unchecked")
+		PersistentTimerHeap(int capacity, ColumnFamilyHandle handle) {
+			this.handle = handle;
+
+			this.partitionHeadTimers = (Tuple4<Integer, Long, K, N>[])new Tuple4[capacity];
+			this.indices = new int[capacity];
+			this.orders = new int[capacity];
+			
+			for (int i = 0; i < capacity; i++) {
+				orders[i] = i;
+				indices[i] = i;
+			}
+		}
+
+		private int getParentOrder(int order) {
+			return (order - 1) / 2;
+		}
+
+		private int getLeftChildOrder(int order) {
+			return order * 2 + 1;
+		}
+
+		private int getRightChildOrder(int order) {
+			return order * 2 + 2;
+		}
+
+		private long getTimestamp(Tuple4<Integer, Long, K, N> rawTimer) {
+			return rawTimer == null ? Long.MAX_VALUE : rawTimer.f1;
+		}
+
+		/** Rebuild the heap with the given DB. */
+		void initialize() {
+			int currentPartition = -1;
+			Tuple4<Integer, Long, K, N> currentPartitionLeaderTimer = null;
+
+			RocksIterator iterator = db.newIterator(handle);
+			iterator.seekToFirst();
+
+			while (iterator.isValid()) {
+				Tuple4<Integer, Long, K, N> rawTimer = deserializeRawTimer(iterator.key());
+
+				int partition = getPartitionForKeyGroup(keyGroupRange, rawTimer.f0, numPartitions);
+				if (partition == currentPartition) {
+					if (currentPartitionLeaderTimer == null || rawTimer.f1 < currentPartitionLeaderTimer.f1) {
+						currentPartitionLeaderTimer = rawTimer;
+					}
+				} else {
+					if (currentPartitionLeaderTimer != null) {
+						updatePartitionLeader(currentPartition, currentPartitionLeaderTimer);
+					}
+
+					currentPartition = partition;
+					currentPartitionLeaderTimer = rawTimer;
+				}
+
+				int nextKeyGroup = rawTimer.f0 + 1;
```

Comment:
For example, here it would be more beneficial to have a dedicate class instead of a `Tuple4` for the timers. Like this, it is harder to figure out the meanings of f0, f1, etc.